File: plugins/server/bslScheduler.js
/*
    Copyright © 2011, 2012 MLstate

    This file is part of OPA.

    OPA is free software: you can redistribute it and/or modify it under the
    terms of the GNU Affero General Public License, version 3, as published by
    the Free Software Foundation.

    OPA is distributed in the hope that it will be useful, but WITHOUT ANY
    WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
    FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for
    more details.

    You should have received a copy of the GNU Affero General Public License
    along with OPA.  If not, see <http://www.gnu.org/licenses/>.
*/
##extern-type continuation('a)
// '
##extern-type Scheduler.key

##register sleep : int, (-> void) -> void
##args(delay,todo)
{
    setTimeout(function(){ todo() },delay);
    return js_void;
}

##register asleep : int, (-> void) -> Scheduler.key
##args(delay, todo)
{
    var key = setTimeout(function(){ todo() }, delay);
    return key;
}

##register push : (-> void) -> void
##args(todo)
{
    // WHAT ?
    #<Ifstatic:OPA_CPS_CLIENT>
    push(new Task_from_application(todo, []));
    return js_void;
    #<Else>
    return %%BslScheduler.sleep%%(0, todo);
    #<End>
}

##register timer : int, (-> void) -> void
##args(delay,todo)
{
    function js_todo(){
        todo();
        setTimeout(js_todo,delay);
    }
    setTimeout(js_todo,delay);
    return js_void;
}

##register abort : Scheduler.key -> void
##args(key)
{
    clearTimeout(key);
    return js_void;
}

##register [cps-bypass] sleep_cps : \
int, (continuation(opa[void]) -> void), continuation(opa[void]) -> void
##args(delay, todo, k)
{
    %%BslScheduler.sleep%%(delay, opa_cps_callback_to_js_callback0(k, todo));
    return_(k, js_void);
}

##register [cps-bypass] asleep_cps : \
int, (continuation(opa[void]) -> void), continuation(opa[Scheduler.key]) -> void
##args(delay, todo, k)
{
    var key = %%BslScheduler.asleep%%(delay, opa_cps_callback_to_js_callback0(k, todo));
    return_(k, key);
}

##register [cps-bypass] push_cps : \
(continuation(opa[void]) -> void), continuation(opa[void]) -> void
##args(todo, k)
{
    %%BslScheduler.push%%(opa_cps_callback_to_js_callback0(k, todo));
    return_(k, js_void);
}

##register [cps-bypass] timer_cps :                     \
int, (continuation(opa[void]) -> void), continuation(opa[void]) -> void
##args(delay, todo, k)
{
    %%BslScheduler.timer%%(delay, opa_cps_callback_to_js_callback0(k, todo));
    return_(k, js_void);
}

// Stub
// TODO Something with the scheduler is probably missing

var tty = require("tty");
process.openStdin().on("keypress", function(_chunk, key) {
  if(key && key.name === "c" && key.ctrl) process.exit(0);
});
tty.setRawMode(true);

process.on('SIGINT', function (){process.exit(1)});
process.on('SIGKILL', function (){process.exit(1)});
process.on('SIGTERM', function (){process.exit(1)});

##register at_exit : (-> void) -> void
##args(a)
{
    process.on('exit', function () {
       a();
    });
}

File: plugins/server/bslScheduler.ml
(*
    Copyright © 2011, 2012 MLstate

    This file is part of OPA.

    OPA is free software: you can redistribute it and/or modify it under the
    terms of the GNU Affero General Public License, version 3, as published by
    the Free Software Foundation.

    OPA is distributed in the hope that it will be useful, but WITHOUT ANY
    WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
    FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for
    more details.

    You should have received a copy of the GNU Affero General Public License
    along with OPA. If not, see <http://www.gnu.org/licenses/>.
*)
(** TODO - plugins dependencies *)
##property[mli]
##extern-type continuation('a) = 'a QmlCpsServerLib.continuation
##property[endmli]
(** *****************************)

##extern-type Scheduler.key = Scheduler.async_key

module BslUtils = OpabslgenMLRuntime.BslUtils

let opa = Scheduler.default

let push f = Scheduler.push opa f

let at_exit f = Scheduler.at_exit opa f

let asleep t f =
  Scheduler.sleep opa (Time.milliseconds t) f

let sleep t f = ignore(asleep t f)

let timer t f = Scheduler.timer opa (Time.milliseconds t) f

let abort a = Scheduler.abort opa a

let set_max_compute_successive n = Scheduler.set_max_compute_successive n

let set_nb_step_apply n = QmlCpsServerLib.set_nb_step_apply n

(* BSL REGISTERING *)
##register sleep : int, (-> void) -> void

##register asleep : int, (-> void) -> Scheduler.key

##register [cps-bypass] sleep_cps : int, (continuation(opa[void]) -> void), continuation(opa[void]) -> void
let sleep_cps t f k =
  sleep t (BslUtils.proj_cps0 k f);
  QmlCpsServerLib.return k ServerLib.void

##register [cps-bypass] asleep_cps : int, (continuation(opa[void]) -> void), continuation(opa[Scheduler.key]) -> void
let asleep_cps t f k =
  QmlCpsServerLib.return k (asleep t (BslUtils.proj_cps0 k f))


##register timer : int, (-> void) -> void

##register [cps-bypass] timer_cps : int, (continuation(opa[void]) -> void), continuation(opa[void]) -> void
let timer_cps t f k =
  timer t (BslUtils.proj_cps0 k f);
  QmlCpsServerLib.return k ServerLib.void


##register push : (-> void) -> void

##register [cps-bypass] push_cps : (continuation(opa[void]) -> void), continuation(opa[void]) -> void
let push_cps f k =
  push (BslUtils.proj_cps0 k f);
  QmlCpsServerLib.return k ServerLib.void

##register at_exit : (-> void) -> void

##register [cps-bypass] at_exit_cps : (continuation(opa[void]) -> void), continuation(opa[void]) -> void
let at_exit_cps f k =
  at_exit (BslUtils.proj_cps0 k f);
  QmlCpsServerLib.return k ServerLib.void

##register finalize : ('a -> void), 'a -> void
let finalize f v = Scheduler.finalise opa f v

##register [cps-bypass] finalize_cps : (opa['a], continuation(opa[void]) -> void), opa['a], continuation(opa[void]) -> void
let finalize_cps f v k =
  finalize (BslUtils.proj_cps k f) v;
  QmlCpsServerLib.return k ServerLib.void

##register set_max_compute_successive : int -> void

##register set_nb_step_apply : int -> void
